pyart/retrieve/kdp_proc.py

```diff
@@ -0,0 +1,699 @@
+    if debug:
+        unique = [len(slc) for slc in slices if slc is not None]
+        print 'Total number of unique regions: {}'.format(sum(unique))
+
+    phi_near = np.zeros(radar.nrays, dtype=psidp.dtype)
+    phi_far = np.zeros(radar.nrays, dtype=psidp.dtype)
+    range_near = np.zeros(radar.nrays, dtype=psidp.dtype)
+    range_far = np.zeros(radar.nrays, dtype=psidp.dtype)
+    idx_near = np.zeros(radar.nrays, dtype=np.int32)
+    idx_far = np.zeros(radar.nrays, dtype=np.int32)
+
+    for ray, regions in enumerate(slices):
+
+        # check if all range gates are missing
+        if regions is None:
+            continue
+
+        else:
+
+            # near range gate boundary condition
+            for slc in regions:
+
+                # check if enough samples exist in slice
+                if slc.stop - slc.start >= n:
+
+                    # parse range gate indices and nearest range gate
+                    idx = slice(slc.start, slc.start + n)
+                    idx_near[ray] = idx.start
+                    range_near[ray] = radar.range['data'][idx][0]
+
+                    # parse data for linear regression and compute slope
+                    x = radar.range['data'][idx]
+                    y = psidp[ray,idx]
+                    slope = stats.linregress(x, y)[0]
+
+                    # if linear regression slope is positive, set near range
+                    # gate boundary condition to first differential phase
+                    # measurement, otherwise use the median value
+                    if slope > 0.0:
+                        phi_near[ray] = y[0]
+                    else:
+                        phi_near[ray] = np.median(y)
+                    break
+
+                else:
+                    continue
+
+            # far range gate boundary condition
+            for slc in reversed(regions):
+
+                if slc.stop - slc.start >= n:
+
+                    # parse index of furthest range gate
+                    # parse range gate indices and furthest range gate
+                    idx = slice(slc.stop - n, slc.stop)
+                    idx_far[ray] = idx.stop
+                    range_far[ray] = radar.range['data'][idx][-1]
+
+                    # parse data for linear regression and compute slope
+                    x = radar.range['data'][idx]
+                    y = psidp[ray,idx]
+                    slope = stats.linregress(x, y)[0]
+
+                    # if linear regression slope is positive, set far range
+                    # gate boundary condition to last differential phase
+                    # measurement, otherwise use the median value
+                    if slope > 0.0:
+                        phi_far[ray] = y[-1]
+                    else:
+                        phi_far[ray] = np.median(y)
+                    break
+
+                else:
+                    continue
+
+    # check for outliers in the near range gate boundary conditions
```

Comment:
This is the section in `boundary_conditions_maesaka` that will address outliers in PHIDP_near. Additional pushes to this PR will address this (hopefully).